### ISSUE TITLE
Help with MAKE IT MODULAR..

### ISSUE THREAD

**Issue by theCard** [nodeschool/discussions#64]
I can't figure out what I am doing wrong. My ACTUAL matches my EXPECTED in the output, but I am getting this error...

Your additional module file:
P:\NODEJS1\chapter02\mymodule.js
did not return the correct list of files in an array as the second argument of the callback.

Here is my program...

```
var mymodule = require('./mymodule.js')

var dir = process.argv[2]
var ext = process.argv[3]

mymodule(dir, ext, function(err, list) {
  if (err) {
   throw err;
  }
  else list.forEach(function (file) {
   console.log(file);
  })
});
```

AND THE MODULE...

```
var fs = require('fs')

module.exports = function(dirname, filter, callback) {
var regex = new RegExp('\\.' + filter + '$')    

var filelist = new Array()
var i = 0

fs.readdir(dirname, function (err, list) {

    if (err) {
    return callback(err);
    }
    else list.forEach(function (file) {
    if (regex.test(file))
        filelist[i] = file;
        i += 1;
    })
    return callback(null, filelist)

})

};
```

**Comment by WADaniel** [comment]
Here's my solution:

program.js

```javascript
var filter  = require('./module.js');
var dir     = process.argv[2];
var ext     = process.argv[3];

filter(dir, ext, function(err, filteredList) {

    for (var i = 0; i < filteredList.length; i++) {
        console.log(filteredList[i]);
    }

});
´´´

module.js

```javascript
module.exports = function (dir, ext, callback) {

    var fs   = require('fs');

    fs.readdir(dir, function (err, list) {

        if(err) {

            return callback(err);

        } else {

            var filteredList = list.filter(sort);

            function sort(value) {
                if(value.indexOf('.' + ext) >= 0) {
                    return value
                }
            }

        }

        callback(null, filteredList);
    });

};
´´´

**Comment by i-nata** [comment]
program.js


      var filterFn = require('./mymodule.js');
      var dir = process.argv[2]
      var filterStr = process.argv[3]

      filterFn(dir, filterStr, function (err, list) {
        if (err)
           return console.error('There was an error:', err)

       list.forEach(function (file) {
            console.log(file)
        })
      })

mymodule.js


    var fs = require('fs')
    var path = require('path')

    module.exports = function (dir, filterStr, callback) {

     fs.readdir(dir, function (err, list) {
        if (err)
            return callback(err)

        list = list.filter(function (file) {
            return path.extname(file) === '.' + filterStr
        })

        callback(null, list)
     })
    }

**Comment by yuezhizizhang** [comment]
It's a broken test. It asks us to compare the ext string passed into the module directly, but actually, you have to append '.' to the ext to do the comparison. Otherwise, the result complains that the module did not return an Array with the correct number of elements.

    module.exports = function(dir, ext, callback) {
       var fs = require('fs'),
           path = require('path'),
           ext = '.' + ext,
           list = [];

       fs.readdir(dir, function(err, data) {
          if (!!err) {
             return callback(err, null);
          }

          list = data.filter(function(file) {
             return path.extname(file) === ext;
          });

          return callback(null, list);
       });
    };

**Comment by martinheidegger** [comment]
This thread is too long. Pulling in @rvagg: Do you think its necessary to torture readers by asking them for this tricky part: It seems like many people dislike this check.
Basically its semantics, if we were to write: `The extension will be without a prepending "."!` and in the test  `Your solution seems to not check the prepending "."` Maybe that would be easier to understand.

**Comment by rvagg** [comment]
yes, it seems to be torture for many, but what I like about this exercise is that you have to build an API to a specification and you can't deviate from that.

however, in the interests of cutting down on the endless issues that get filed on this repo, we could probably integrate a more verbose check and report explicitly what they are doing wrong. Who wants to do that work?

**Comment by JVimes** [comment]
This Issue is closed, but even the latest incarnation of the exercise is more of a time-sink than I think it should be. I spent an extra hour or two trying to figure out what result format would satisfy the validator, even though it said my values were correct. I'm trying to learn Modules, not practice interpreting an API spec written in prose. I have limited time, so it's frustrating.

**Comment by martinheidegger** [comment]
@JVimes Sharing the pain. The work on the improvement is ongoing. Please know that usability issues are issues worth writing about.

**Comment by JVimes** [comment]
@martinheidegger, thanks for understanding! Really appreciate it.

**Comment by lahsrah** [comment]
Any idea why mine doesn't work?

Program:
    var dir = process.argv[2];
var extWithoutDot = process.argv[3];
var extWithDot = "." + extWithoutDot;

var program6module = require('./program6_module.js');

program6module(dir, extWithDot, afterRead);

function afterRead(err, data) {
	if(err)
		console.log(err);
	else 
	{
		data.forEach(function(params) {
			console.log(params);
		})
	}
}


Module:

var fs = require('fs');
var path = require('path');

module.exports = function filterDir(dir, extension, callback) {
		fs.readdir(dir, function(err, fileNameArray){
			
			if(err) 
			{
				callback(err, null);
			} else {
			
				var filtered = filter(fileNameArray, extension);
				console.log(filtered);
				callback(null, filtered);
			}
		});
};

function filter(fileNameArray, extension){
	return fileNameArray.filter(function(x){
		return path.extname(x) === extension;
	})
}

**Comment by CharlieStras** [comment]
I think the error message isn't correct, most of the problem happened as we added the dot outside the module. So why the error message say something about the second argument of the callback?

**Comment by tttmaximttt** [comment]
```var patt = new RegExp("\\." + process.argv[3] + "$", "i")``` - you have excess ```\``` symbol
```var patt = new RegExp("\." + process.argv[3] + "$", "i")``` - this is right;

**Comment by tttmaximttt** [comment]
```
                 ACTUAL                                 EXPECTED
────────────────────────────────────────────────────────────────────────────────

   "CHANGELOG.md"                      ==    "CHANGELOG.md"
   "LICENCE.md"                        ==    "LICENCE.md"
   "README.md"                         ==    "README.md"
   ""                                  ==    ""

────────────────────────────────────────────────────────────────────────────────

✓ Submission results match expected
✓ Additional module file exports a single function
✓ Additional module file exports a function that takes 3 arguments
✓ Additional module file handles errors properly
✓ Additional module file handles callback argument
✓ Additional module file returned two arguments on the callback function
✓ Additional module file returned correct number of elements as the second argument of the callback
✗ Your additional module file [program.js] may be matching "ext" instead of ".ext"

# FAIL

Your solution to MAKE IT MODULAR didn't pass. Try again!```
Who can help me, what I'm doing wrong!

**Comment by tttmaximttt** [comment]
Who can help me, what I'm doing wrong!

**Comment by smrutimandal** [comment]
Thanks so @kmerhi . You made my day. :)

**Comment by blueoasis** [comment]
Here's the official solution in case you want to compare notes:

────────────────────────────────────────────────────────────────────────────────
solution.js:

    var filterFn = require('./solution_filter.js')
    var dir = process.argv[2]
    var filterStr = process.argv[3]

    filterFn(dir, filterStr, function (err, list) {
      if (err)
        return console.error('There was an error:', err)

      list.forEach(function (file) {
        console.log(file)
      })
    })

────────────────────────────────────────────────────────────────────────────────
solution_filter.js:

    var fs = require('fs')
    var path = require('path')

    module.exports = function (dir, filterStr, callback) {

      fs.readdir(dir, function (err, list) {
        if (err)
          return callback(err)

        list = list.filter(function (file) {
          return path.extname(file) === '.' + filterStr
        })

        callback(null, list)
      })
    }

For me, the issue was that I was not executing the callback(null, list) in case of success, and I was not returning callback(err) in case of an error. My program produced the correct results, but did not adhere to these rules. The main issue, as I see it, is that this is not clearly stated in the program requirements, but only after validating the program.

**Comment by pavelcreates** [comment]
Would anyone be able to help me out by explaining why my results come out crazy?

**mymodule.js**
```javascript
function myModule(file, filter, callback) {
	
	var fs = require('fs'),
		path = require('path'),
		n = 0,
		results = [];

	fs.readdir(file, function(err, data) {
		if (err) {
			return callback(err);
		}

		for (i = 0; i < data.length; i++) {

			if (path.extname(data[i]) === "." + filter) {
				results[n] = data[i];
				n++;
			}

			callback(null, results);
		}

	});

}

module.exports = myModule;
```



**makeitmodular.js**
```javascript
var myModule = require('./mymodule'),
	file = process.argv[2],
	filter = process.argv[3];

myModule(file, filter, function(err, result) {
	for (var i = 0; i < result.length; i++) {
		console.log(result[i]);
	}
})
```

My results come out as so:
```
CHANGELOG.md
CHANGELOG.md
LICENCE.md
CHANGELOG.md
LICENCE.md
README.md
CHANGELOG.md
LICENCE.md
README.md
CHANGELOG.md
LICENCE.md
README.md
CHANGELOG.md
LICENCE.md
README.md
CHANGELOG.md
LICENCE.md
README.md
// continues for a while in the same pattern - I shortened it to reduce the spam
```

**Comment by nikitasingh** [comment]
@pavelcreates you are calling the callback inside for loop. Take it out of the loop and and it will work. :)

**Comment by krizka974** [comment]
I got an 'undefined' line that I don't understand here. Someone can help?

make-it-modular.js
```javascript
var fs = require('fs')
var path = require('path')

module.exports = function (dirPath, fileExt, callback){
fs.readdir(dirPath, function (err, list){
	if(err) {
		return callback(err)
	} else {
		var newList = []
		for (i=0;i<list.length;i++){
			if(path.extname(list[i]) === '.'+fileExt){
				newList.push(list[i]);
			}
		}
		callback(null, newList)
	}
})};
```

make-it-modular-2.js
```javascript
var myModule = require('./make-it-modular.js')

var dirPath = process.argv[2]
var fileExt = process.argv[3]

console.log(myModule(dirPath,fileExt,function(err,res){
	if(err) return err;
	res.forEach(function(file){
		console.log(file)
	})
}));
```

The result

>Your submission results compared to the expected:
>
>                 ACTUAL                                 EXPECTED
>────────────────────────────────────────────────────────────────────────────────
>
>   "undefined"                         !=    "CHANGELOG.md"
>   "CHANGELOG.md"                      !=    "LICENCE.md"
>   "LICENCE.md"                        !=    "README.md"
>   "README.md"                         !=    ""
>   ""                                  !=
>
>────────────────────────────────────────────────────────────────────────────────
>
> ✗ Submission results did not match expected!
> ✓ Additional module file exports a single function
> ✓ Additional module file exports a function that takes 3 arguments
> ✓ Additional module file handles errors properly
> ✓ Additional module file handles callback argument
> ✓ Additional module file returned two arguments on the callback function
> ✓ Additional module file returned correct number of elements as the second argument of the callback
> ✓ Additional module file correctly handles '.'-prefixed extension
> ✓ Additional module file returned correct number of elements as the second argument of the callback
> ✓ Additional module file returned correct list of files as the second argument of the callback
>
>     # FAIL

**Comment by MykonBlu** [comment]
@krizka974 The extra undefined is because you are using `console.log` on your entire `myModule` call.

**Comment by anmolview** [comment]
#Pass

program6.js

var dir = process.argv[2];
var format = process.argv[3];
var ls= require('./listfile.js');

ls(dir, format, function (err, data)
{
	if(err)
	{
		console.log('error : '+err);
	}
	for(i in data)
	{
			console.log(data[i]);
	}
});


listfile.js

var dir = process.argv[2];
var format = process.argv[3];
var ls= require('./listfile.js');

ls(dir, format, function (err, data)
{
	if(err)
	{
		console.log('error : '+err);
	}

	for(i in data)
	{
		console.log(data[i]);
	}
});

**Comment by finnp** [comment]
@anmolview Please [open a new issue](https://github.com/nodeschool/discussions/issues/new) with your problem. This way it's easier for people to help you with your problem. This issue has been closed.
